std/protocols/permutation.asm

```diff
@@ -1,44 +1,152 @@
-let permutation = |acc, lhs_selector, lhs, rhs_selector, rhs| {
+// - acc: A phase-2 witness column to be used as the accumulator. If 2 are provided, computations
+//        are done on the F_{p^2} extension field.
+// - permutation_constraint: The permutation constraint
+let permutation: expr[], Constr -> Constr[] = |acc, permutation_constraint| {
+
+    let (lhs_selector, lhs, rhs_selector, rhs) = unpack_permutation_constraint(permutation_constraint);
 
     let _ = assert(len(lhs) == len(rhs), || "LHS and RHS should have equal length");
-    let _ = assert(modulus() > 2**100, || "This implementation assumes a large field");
+    let with_extension = match len(acc) {
+        1 => false,
+        2 => true,
+        _ => panic("Expected 1 or 2 accumulator columns!")
+    };
+
+    let _ = if !with_extension {
+        assert(!needs_extension(), || "The Goldilocks field is too small and needs to move to the extension field. Pass two accumulators instead!")
+    } else { [] };
+
+    // On the extension field, we'll need two field elements to represent the challenge.
+    // If we don't need an extension field, we can simply set the second component to 0,
+    // in which case the operations below effectively only operate on the first component.
+    let fp2_from_array = |arr| if with_extension { Fp2::Fp2(arr[0], arr[1]) } else { from_base(arr[0]) };
+    let acc_ext = fp2_from_array(acc);
+    let alpha = fp2_from_array([alpha1, alpha2]);
+    let beta = fp2_from_array([beta1, beta2]);
 
     // If the selector is 1, contribute a factor of `beta - compress_expression_array(lhs)` to accumulator.
-    // If the selector is 0, contribute a factor of 1 to the accumulator. 
-    let lhs_folded = lhs_selector * (beta - compress_expression_array(lhs) - 1) + 1;
-    let rhs_folded = rhs_selector * (beta - compress_expression_array(rhs) - 1) + 1;
+    // If the selector is 0, contribute a factor of 1 to the accumulator.
+    // Implemented as: folded = selector * (beta - compress_expression_array(values) - 1) + 1;
+    let lhs_folded = selected_or_one(lhs_selector, sub_ext(beta, compress_expression_array(lhs, alpha)));
+    let rhs_folded = selected_or_one(rhs_selector, sub_ext(beta, compress_expression_array(rhs, alpha)));
+
+    let next_acc = if with_extension {
+        next_ext(acc_ext)
+    } else {
+        // The second component is 0, but the next operator is not defined on it...
+        from_base(acc[0]')
+    };
+
+    // Update rule:
+    // acc' = acc * lhs_folded / rhs_folded
+    // => rhs_folded * acc' - lhs_folded * acc = 0
+    let (update_expr_1, update_expr_2) = unpack_ext(sub_ext(
+        mul_ext(rhs_folded, next_acc),
+        mul_ext(lhs_folded, acc_ext)
+    ));
+
+    let (acc_1, acc_2) = unpack_ext(acc_ext);
 
     [
         // First and last z needs to be 1
         // (because of wrapping, the z[0] and z[N] are the same)
-        is_first * (acc - 1) = 0,
+        is_first * (acc_1 - 1) = 0,
```

Comment:
Yeah, here we are at the point where we are implementing a logical implication using polynomial constraints. If we had a higher-level concept for that, we could do that generically for both the extension field case and the regular field case.

`implies(is_first, equal(acc, 1))`

Or even

`implies(is_first, acc = 1)`

if we have operators.